typescript/schedules.md

```diff
@@ -0,0 +1,1147 @@
+   */
+  connection?: ConnectionLike;
+
+  /**
+   * Server namespace
+   *
+   * @default default
+   */
+  namespace?: string;
+}
+
+export type ScheduleClientOptionsWithDefaults = Replace<
+  Required<ScheduleClientOptions>,
+  {
+    connection?: ConnectionLike;
+  }
+>;
+export type LoadedScheduleClientOptions = ScheduleClientOptionsWithDefaults & {
+  loadedDataConverter: LoadedDataConverter;
+};
+
+export function defaultScheduleClientOptions(): ScheduleClientOptionsWithDefaults {
+  return {
+    dataConverter: {},
+    // The equivalent in Java is ManagementFactory.getRuntimeMXBean().getName()
+    identity: `${process.pid}@${os.hostname()}`,
+    interceptors: {},
+    namespace: 'default',
+  };
+}
+
+/**
+ * Policy for overlapping Actions.
+ */
+export enum ScheduleOverlapPolicy {
+  /**
+   * Use server default (currently SKIP).
+   *
+   * TODO remove this field if this issue is implemented: https://github.com/temporalio/temporal/issues/3240
+   */
+  UNSPECIFIED = 0,
+
+  /**
+   * Don't start a new Action.
+   */
+  SKIP,
+
+  /**
+   * Start another Action as soon as the current Action completes, but only buffer one Action in this way. If another
+   * Action is supposed to start, but one Action is running and one is already buffered, then only the buffered one will
+   * be started after the running Action finishes.
+   */
+  BUFFER_ONE,
+
+  /**
+   * Allows an unlimited number of Actions to buffer. They are started sequentially.
+   */
+  BUFFER_ALL,
+
+  /**
+   * Cancels the running Action, and then starts the new Action once the cancelled one completes.
+   */
+  CANCEL_OTHER,
+
+  /**
+   * Terminate the running Action and start the new Action immediately.
+   */
+  TERMINATE_OTHER,
+
+  /**
+   * Allow any number of Actions to start immediately.
+   *
+   * This is the only policy under which multiple Actions can run concurrently.
+   */
+  ALLOW_ALL,
+}
+
+checkExtendsWithoutPrefix<
+  temporal.api.enums.v1.schedule.ScheduleOverlapPolicy,
+  ScheduleOverlapPolicy,
+  'SCHEDULE_OVERLAP_POLICY_'
+>();
+
+export interface Backfill {
+  /** Time range to evaluate Schedule in. */
+  start: Date;
+  end: Date;
+
+  /**
+   * Override the Overlap Policy for this request.
+   */
+  overlap?: ScheduleOverlapPolicy;
+}
+
+/**
+ * The range of values depends on which, if any, fields are set:
+ *
+ * ```
+ * {} -> all values
+ * {start} -> start
+ * {start, end} -> every value from start to end (implies step = 1)
+ * {start, step} -> from start, by step (implies end is max for this field)
+ * {start, end, step} -> from start to end, by step
+ * {step} -> all values, by step (implies start and end are full range for this field)
+ * {end} and {end, step} are not allowed
+ * TODO is there a way to express not allowed ^ in type?
+ * ```
+ *
+ * For example:
+ *
+ * ```
+ * {start: 2, end: 10, step: 3} -> 2, 5, 8
+ * ```
+ */
+export interface Range<Unit> {
+  start?: Unit;
+  end?: Unit;
```

Comment:
```suggestion
 * {} -> all values
 * {start} -> start
 * {start, end} -> every value from start to end
 * {start, end, step} -> from start to end, by step
 * {step} -> all values, by step (implies start and end are full range for this field)
 * {end} and {end, step} are not allowed
 * TODO is there a way to express not allowed ^ in type?
 * ```
 *
 * For example:
 *
 * ```
 * {start: 2, end: 10, step: 3} -> 2, 5, 8
 * ```
 */
export interface Range<Unit> {
  start?: Unit;
  
  /**
   * @default start
   */
  end?: Unit;
```

@dnr does the below work?

```
 * {step} -> all values, by step (implies start and end are full range for this field)
```

Reply:
no, that means start == end == 0 (default values for proto messages), which means it will match the single value 0, regardless of step